longitudinal_controller/src/longitudinal_controller.py

```diff
@@ -0,0 +1,227 @@
+#!/usr/bin/python3
+
+from enum import Enum
+import numpy as np
+import rospy
+import time
+
+from pid import PID
+from geometry_msgs.msg import TwistStamped
+from self_racing_car_msgs.msg import ArduinoLogging
+from std_msgs.msg import Float32, Float64MultiArray
+
+
+class LongitudinalControlMode(Enum):
+    ConstantPwmOutput = 1
+    PID = 2
+
+
+class LongitudinalController:
+    def __init__(self):
+
+        # Constants
+        rate = rospy.get_param("~rate", 10.0)
+        self.throttle_idle_autonomous_pwm = rospy.get_param(
+            "~throttle_idle_autonomous_pwm", 90
+        )
+        self.throttle_max_autonomous_pwm = rospy.get_param(
+            "~throttle_max_autonomous_pwm", 102
+        )
+        self.throttle_min_autonomous_pwm = rospy.get_param(
+            "~throttle_min_autonomous_pwm", 70
+        )
+        self.longitudinal_control_mode = rospy.get_param(
+            "~longitudinal_control_mode", 2
+        )
+        gain_p = rospy.get_param("~speed_control_gain_p", 0.5)
+        gain_i = rospy.get_param("~speed_control_gain_i", 0.1)
+        gain_d = rospy.get_param("~speed_control_gain_d", 0)
+        self.constant_pwm_output = rospy.get_param(
+            "~constant_pwm_output", 90
+        )  # If ConstantPwmOutput mode
+        self.timeout_engage_msg_before_stop_secs = rospy.get_param(
+            "~timeout_engage_msg_before_stop_secs", 1
+        )
+
+        # Subscribers
+        rospy.Subscriber(
+            "target_velocity",
+            TwistStamped,
+            self.target_velocity_callback,
+        )
+        rospy.Subscriber(
+            "current_velocity",
+            TwistStamped,
+            self.current_velocity_callback,
+        )
+        rospy.Subscriber(
+            "arduino_logging",
+            ArduinoLogging,
+            self.firmware_log_callback,
+        )
+
+        # Publishers
+        self.throttle_cmd_pub = rospy.Publisher(
+            "throttle_pwm_cmd",
+            Float32,
+            queue_size=10,
+        )
+
+        self.pub_debug_pid = rospy.Publisher(
+            "pid_debug", Float64MultiArray, queue_size=10
```

Comment:
I'd be down to create a custom message type for this, that way we don't need to remember which field of the message corresponds to which value.

Reply:
Done